Save results directly in google sheets

mandre opened the pull request:
This PR brings a few improvements to gazelle:
- it saves results directly in google sheets
- it pulls the all the latest job results from prow, unless specified with the `-id` option.
- it knows at which position to update or insert rows

Closes #9

mandre committed:
WIP Outputs to google sheets

This commit makes gazelle save the reports to google sheets directly,
without the need to insert lines and paste into the document.

Possible improvement:
- scrap the ID of the latest jobs from prow directly
- insert or replace rows to the correct line based on job IDs, if for
  example running gazelle with a discontinuous range
- calling gazelle without job name inserts latest results for all jobs

mandre committed:
Gofmt the sheet

mandre committed:
Create a gsheets.Client type

mandre committed:
Teach gazelle to figure job IDs

When not specifying job IDs with the `--id` option, gazelle is now able
to get the upper boundary from Prow (excluding pending jobs) and the
lower one from the latest spreadsheet entry.

This is a step further from allowing running gazelle without any
argument.

mandre committed:
Refactoring: Move code into Job

We now pass a job to gsheets.AddRow(), this will allow to easily get
the job ID, in order to insert a row at the right position in the
spreadsheet rather than at the top.

mandre committed:
Fix OKD spreadsheet names

mandre committed:
Introduce the gsheets.Sheet type

mandre committed:
Allow inserting rows at their right position

Rows are now places at their position based on the job ID instead of
always at the top.
This is needed if we run or if we want to update existing rows.

mandre committed:
Update documentation

mandre commented:
This is currently saving into a copy of the CI spreadsheet for testing. Let me know if you're fine merging this PR and I'll update the address of the spreadsheet.

pierreprinetti commented:
> update the address of the spreadsheet

Might be interesting to have this and other config in `$XDG_CONFIG_USER`

pierreprinetti commented on pkg/job/job.go:
```diff
@@ -48,12 +51,12 @@ func (j *Job) Name() (string, error) {
 	f, err := j.fetch(j.baseURL() + "/started.json")
 	if err != nil {
-		return time.Time{}, err
+		panic(err)
```
why panicking here? I don't see it recovered in the package

pierreprinetti commented on pkg/job/job.go:
```diff
@@ -73,6 +76,20 @@ func (j Job) FinishTime() (time.Time, error) {
 	return finished.time, nil
 }
 
+func (j Job) Duration() string {
```
could return a `time.Duration`; let's let the caller use its `String()` method

pierreprinetti commented on pkg/job/job.go:
```diff
@@ -175,3 +192,22 @@ func (j Job) JUnit() (io.Reader, error) {
 	return j.fetch(u)
 }
+
+func (j Job) ToHtml() string {
```
you can call this `HTML()`, if you want to mimic `String()`

pierreprinetti replied:
(more on this method hereafter)

pierreprinetti commented on pkg/job/job.go:
```diff
@@ -15,9 +16,11 @@ import (
+	FullName       string
+	ID             string
+	ComputedResult string
+	UserName       string
```
Having the username here looks like an interface segregation violation to me.

Can we have the `ToHTML` method accept the username as an argument instead?

pierreprinetti commented on pkg/job/job.go:
```diff
@@ -175,3 +192,22 @@ func (j Job) JUnit() (io.Reader, error) {
 }
+
+func (j Job) ToHtml() string {
+	startTime, _ := j.StartTime()
```
ah, now I see why you're panicking in StartTime :joy: 

May I suggest that you either:
* handle service degradation (thus ignoring the error, but without panicking); OR
* guarantee the result and bubble up the error?

pierreprinetti replied:
worst case, you can panic _here_

pierreprinetti commented on pkg/gsheets/gsheets.go:
```diff
@@ -0,0 +1,106 @@
+	spreadsheetId string
+}
+
+func NewClient() Client {
```
Opinion:

This function does I/O: it has all rights to return an error.

pierreprinetti commented on pkg/gsheets/gsheets.go:
```diff
@@ -0,0 +1,106 @@
+		log.Fatalf("Unable to cache oauth token: %v", err)
+	}
+	defer f.Close()
+	json.NewEncoder(f).Encode(token)
```
you may want to catch the error here

pierreprinetti commented on pkg/gsheets/sheet.go:
```diff
@@ -0,0 +1,185 @@
+	return id
+}
+
+func (s *Sheet) AddRow(j job.Job) {
```
Watch out, that's a circular reference:
* `gsheets` knows about `job`, because it needs both job-id and job-html;
* `job` knows about the spreadsheet: the XML logic of `ToHtml` intrinsically does.

You could move `ToHtml` to this package. That way, `job` stays focused on the CI domain; while here we deal with the spreadsheet.

As a side effect, this solution fixes your `Username` problem: it's fine for `gsheets` to know about it.

pierreprinetti replied:
For bonus points, apply dependency inversion the Go way!

You can give `job.Job` a bunch of exported methods, and then consume them through an interface in `gsheet`

pierreprinetti commented:
amazing work!

Just a couple comments regarding the dependency tree mostly; in perspective, if we want to grow this thing we need to keep `job` as independent as possible.

pierreprinetti commented on pkg/gsheets/sheet.go:
```diff
@@ -0,0 +1,185 @@
+		s.idxList[idx] = id
+	}
+
+	pdr := &sheets.PasteDataRequest{
```
wow, congrats for finding this

mandre replied:
Sadly, there's no other way to add rendered HTML into gsheets.

pierreprinetti commented on pkg/gsheets/sheet.go:
```diff
@@ -0,0 +1,185 @@
+	return sheetName
+}
+
+func (s *Sheet) getAllIDs() {
```
why do we need all the IDs? Is that for performing insertions in the middle? Can't we just limit ourselves to appending to the top?

mandre replied:
For a two reasons: you can now set the range in whatever order suits you, 405-407 or 407-405, and you'll get consistent result in the sheet, or pass gazelle the ID of old jobs you want to refresh the results.

pierreprinetti replied:
wouldn't be cool if we started from the top :P

mandre committed:
Address review comments

mandre committed:
Allow running for all jobs if `-job` param is unset

With this, updating the spreadsheet with latest results for all jobs is
as simple as running `./cireport`.

pierreprinetti commented on cmd/cireport/main.go:
```diff
@@ -20,71 +20,89 @@ var (
+			if lowerBound < upperBound {
+				realJobIDs = fmt.Sprintf("%d-%d", lowerBound, upperBound)
+			} else if lowerBound == upperBound {
+				realJobIDs = fmt.Sprintf("%d", upperBound)
```
We should be [covered][1] already

[1]: https://github.com/pierreprinetti/go-sequence/blob/350b74325d6470e42f48e089d1a2ead45bab71a3/sequence_test.go#L93-L99

pierreprinetti reviewed (dismissed):
Seeing it in action is hilarious :grinning: 

Great work!

(lmk when I can approve again, with the actual spreadsheet)

mandre committed:
Use the real CI spreadsheet

pierreprinetti reviewed (dismissed):
:tada:
